core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java

```diff
@@ -1310,7 +1311,9 @@ private static List<RexNode> convertOperands(SqlRexContext cx,
       SqlOperandTypeChecker.Consistency consistency, List<RelDataType> types) {
     switch (consistency) {
     case COMPARE:
-      if (SqlTypeUtil.areSameFamily(types)) {
+      if (SqlTypeUtil.areSameFamily(types)
```

Comment:
so this was a problem of missing casts, not of missing optimizations

Reply:
Yes, JavaType and RelDataType are mixed.